### DISCUSSION TITLE
Register and Login a user.

### DISCUSSION THREAD

**Discussion by Aaron-Ochieng** [pocketbase/pocketbase#6044] [Q&A]
How do I insert a new user(which I've parsed from a form) to the default pocketbase users database?? How do I login a user . All the tutorials I've seen revolves around Javascript. I need to achieve this using golang🤷🏾‍♂️

**Comment by xrd** [comment]
You've got a lot of options. 

When you say login a user, do you mean inside a browser context? If so, you'll probably want to review the docs available in the UI. That happens in JavaScript, but you said you want to use golang. It would help to give more details of what you want to do. 

You can definitely just write golang that directly accesses the SQLite database if you want to create a user. 

Or, you could write a hook in golang that permits you to accept a request that comes from you after parsing a form. 

https://pocketbase.io/docs/go-event-hooks/#onserve

It you provide more information about what you want to do, it'll be easier to help you.

**Reply by Aaron-Ochieng** [reply]
How do I insert records after parsing them from the browser?? And how do I authenticate??

```go
    // 
    se.Router.POST("/register", func(e *core.RequestEvent) error {
			    username := e.Request.FormValue("username")
			    email := e.Request.FormValue("email")
			    password := e.Request.FormValue("password")
    
			    fmt.Printf("%s \n %s\n %s\n", username, email, password)
			    return nil
		    })
    
		    se.Router.POST("/login", func(e *core.RequestEvent) error {
			    email := e.Request.FormValue("email")
			    password := e.Request.FormValue("password")
    
			    fmt.Printf("%s \n %s\n", email, password)
			    return nil
		    })
```

**Reply by ganigeorgiev** [reply]
Please spend some time reading the documentation and more specifically https://pocketbase.io/docs/go-records/#create-new-record.

Creating a new user is the same as creating any other record programmatically. For example:

```go
collection, err := e.App.FindCollectionByNameOrId("users")
if err != nil {
    return err
}

record := core.NewRecord(collection)
record.Set("username", username)
record.Set("email", email) // or SetEmail
record.Set("password", password) // or SetPassword
err = e.App.Save(record)
```

For _"how do I authenticate"_ - PocketBase is stateless and doesn't have sessions in the traditional sense. A single user is considered authenticated if it has valid `Authorization: TOKEN` header with the request. For how to regenerate tokens programmatically you can check https://pocketbase.io/docs/go-records/#generating-and-validating-tokens.

Please make sure to read also https://pocketbase.io/docs/client-side-sdks/.
**If you are trying to use PocketBase with htmx or some other similar tool then I don't recommend it**, especially if you are new to web development because PocketBase doesn't have the necessary helpers out of the box for such use cases and you'll end up writing a lot of functionality on your own (e.g. dealing with cookies), which kindof defeat the purpose of using PocketBase in the first place.

**Reply by Aaron-Ochieng** [reply]
Thanks
